source/mutex/src/lib.rs

```diff
@@ -73,12 +100,16 @@ impl<R: ScopedRawMutex, T: ?Sized> BlockingMutex<R, T> {
     /// of the Raw mutex. See [`ScopedRawMutex::with_lock()`]'s documentation for
     /// more details
     pub fn with_lock<U>(&self, f: impl FnOnce(&mut T) -> U) -> U {
-        self.raw.with_lock(|| {
+        let res = self.raw.with_lock(|| {
             let ptr = self.data.get();
             // SAFETY: Raw Mutex proves we have exclusive access to the inner data
             let inner = unsafe { &mut *ptr };
-            f(inner)
-        })
+            catch_unwind(AssertUnwindSafe(|| f(inner)))
```

Comment:
Take it or leave it: it might be nice to have a comment explaining why we `catch_unwind` here so that future readers know that it's in order to ensure that the `with_lock` call is exited normally so that it can unlock itself? Not a huge deal.